Add E2E tests for Filter blocks (Price, Attribute, and Active Filters)

opr opened the pull request:
Closes #4643

<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This PR will introduce E2E tests for the `Filter Products by Price` block.

The changes included are:
- Add a class to the `div` containing the Filter Products by Price block in the editor (`wp-block-woocommerce-price-filter`). Previously there was not one so our testing utils that rely on class names couldn't target the block.
- Add a class to the `div` containing the Filter Products by Attribute block in the editor (`wc-block-attribute-filter`).
- Add [`pptr-testing-library`](https://www.npmjs.com/package/pptr-testing-library) so we can use the queries provided by it.
- Add `clearAndFillInput` util so we can type into input boxes.
- Add attributes (Capacity and Shade) to the E2E test store.
- Create two new products that use the Capacity attribute.
- Ensure attributes are added correctly to the product by ID (see https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4562/files#diff-2fccf9f205d4c46e7393da8019992bb5bd730aa2a3e854420bba4cdf661ee3ffR362-R398) - This creates the attributes in the store then returns their IDs. We can't know ahead of time what the ID will be so it needs to be fetched from the API after creation.
- Add tests for `Active Filters`, `Filter Products by Price`, and `Filter Products by Attribute`.

<!-- Reference any related issues or PRs here -->
See #2395 huge thanks to @senadir for the initial work done in that PR 🙌🏼 

### How to test the changes in this Pull Request:

1. Ensure the Filter Products by Price block works in the editor and on the front-end still.
2. Ensure the Filter Products by Attribute block works in the editor and on the front-end still.
2. Run the E2E tests - verify all pass.
3. Wait for tests to go green here ✅ 

<!-- If you can, add the appropriate labels -->

### Changelog

> Add end-to-end tests to cover the `Active Filters`, `Filter Products by Attribute`, and `Filter Products by Price` blocks.

opr committed:
Add clear and fill input test util

opr committed:
Add tests for Filter Products by Price

opr committed:
Add class name to filter products by price block in editor

opr committed:
Update wp-env

opr committed:
Fix formatting

opr committed:
Update lock

opr committed:
Revert "Update lock"

This reverts commit f4e809e000430eea3972c94fd2aea0a19ba6961b.

opr committed:
Create product attributes during setup

opr committed:
Remove product attributes during teardown

opr committed:
Add attributes in fixture-data

opr committed:
Add two new products, for attribute testing

opr committed:
Load attributes into store and apply them to products

opr committed:
Add class to Filter by Attribute when editing

This is so it can be targeted in E2E tests more easily.

opr committed:
Add tests for Filter Product by Attribute

opr committed:
Fix formatting in attribute-filter.test.js

opr committed:
Add tests for active filters block

opr committed:
Add doc block for ensuteCleanAttributes

opr committed:
Add saveOrPublish test util

ralucaStan commented on tests/e2e/specs/backend/active-filters.test.js:
```diff
@@ -0,0 +1,77 @@
+		await expect( page ).toRenderBlock( block );
+	} );
+
+	it( 'allows title can be manipulated', async () => {
```
This test description could be changed to `allows changing the block's title`

ralucaStan commented on tests/e2e/specs/backend/active-filters.test.js:
```diff
@@ -0,0 +1,77 @@
+	it( 'allows title can be manipulated', async () => {
+		await openDocumentSettingsSidebar();
+		await clearAndFillInput(
+			'.wp-block[data-type="woocommerce/active-filters"] textarea.wc-block-editor-components-title',
```
I love that we now have the pptr-testing-library that allows us to search for content from a user's point of view. With Testing library I know that it's encouraged to search the page for texts the user sees, so I wonder if we could drop searching by this selector and use 

`$document.findByPlaceholderText('Active filters')`

opr replied:
Hmm, I agree it would be _better_, but I'm not confident this is a good time to start using that in place of what we're doing now with selectors. My reasoning is: if we go ahead with rewriting the `clearAndFillInput` method, we will need to add some kind of fallback for when it's not possible to select something by these "text" values from `pptr-testing-library`. This then means we're juggling two things, puppeteer `elementHandles` and dom-testing-library's `HTMLElement`. Also, all tests using this will need to be rewritten to keep track of the current text. This is because some inputs don't have placeholders, or labels, see Filter by Price title, it's a `textarea` with no label or placeholder, so we'll need to know the text beforehand, then keep track of the the text it changes to.

I apologise if I've misunderstood the way nodes/elements are returned by both of these libraries, please correct me if I'm thinking about this wrong.

On top of this, based on what I read here (https://testing-library.com/docs/example-input-event/) we wouldn't be able to simulate typing into an input, which is sometimes required, because changing the `.value` property of a node isn't enough, in the case of the `Filter Products by Price` title, setting `.value` doesn't persist the new title, it has to be typed "manually".

ralucaStan replied:
Maybe I got a bit carried away because I am a huge fan of the [Testing library's principles](https://testing-library.com/docs/), moving away from testing implementation details (aka search for nested selectors) and, instead, test using what the user sees (texts, labels, etc).

It wasn't fully clear to me why `pptr-testing-library` was added. You mentioned making use of its queries and I thought that would be the default way from now on. I think it would be helpful to explain, maybe in a P2:
- what are the reasons behind adding it
- when this library should be used
- how do we see the adoption phase

I did not know about the limitations around typing so thanks for clearing those out. We can keep `clearAndFillInput` as it is.

opr replied:
Yup, it was brought in because I based my work here on Nadir's PR - https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2756

I liked the idea that we didn't have to go and find controls by their css selector, and could use more "user-based" selectors, so I wanted to bring it over to this PR too. In places that selectors had already been used, I wasn't sure if I should go back and rewrite them. I can certainly do that, though.

opr replied:
I've rewritten all the places where it makes sense to do so (in files touched by this PR!) to use the `pptr-testing-library` queries. This only amounted to 6 or 7 code changes, though. Another look over would be appreciated!

ralucaStan commented on tests/e2e/specs/backend/active-filters.test.js:
```diff
@@ -0,0 +1,77 @@
+		);
+	} );
+
+	it( 'Display style to change', async () => {
```
`allows changing the Display Style`?

ralucaStan reviewed (commented):
Congrats on this PR, your efforts will benefit the project greatly. 

I have added some small comments. 
I salute the addition of pptr-testing-library and bringing us closer to some of Testing library's tools. 

I think that moving forward we should also try to test according to React Testing Library's testing principles. I think this PR would be a nice starting point and showcase of using them.

I recommend to try and change in the new tests you've added selecting elements by css selectors and using pptr-testing-library selectors instead (getByText, findByLabelText, etc...) that allow selecting from an user point of view.

opr committed:
Update wording of title test in all filter block E2E

opr committed:
Update test wording to read better

opr committed:
Update snapshots

ralucaStan commented on tests/e2e/specs/backend/active-filters.test.js:
```diff
@@ -0,0 +1,77 @@
+			'Active filters'
+		);
+		await page.click(
+			'.components-toolbar button[aria-label="Heading 3"]'
```
It is not clear what the test is clicking here. Maybe a comment above could explain this step. Also, this would be like a good use case for searching something by text, if possible, as it is more human-friendly.

Also, Heading 3 is not a human-friendly aria-label so this is something that we might need to fix in another ticket

opr replied:
Added a comment!

I don't think we have control over the label though, that comes from Gutenberg. It refers to the control in the block sidebar: 
<img width="250" src="https://user-images.githubusercontent.com/5656702/129210314-637b6836-9d29-4a06-b616-a20eb613b805.png" />

ralucaStan reviewed (commented):
Re-review done

opr committed:
Use pptr-testing-library selectors to click elements

This makes the tests easier to reason with, and lets us take a step back from the implementation details (classnames etc.)

opr committed:
Use pptr-testing-library selectors to click elements in attribute filter

opr committed:
Use pptr-testing-library selectors to click elements in price filter

senadir commented on tests/e2e/fixtures/fixture-loaders.js:
```diff
@@ -334,7 +352,76 @@ const deleteBlockPages = ( ids ) => {
 	);
 };
 
+/**
+ * Ensure attributes are removed from the site before setting up. This exists in case
+ * the teardown didn't happen properly for any reason. This function exists mostly as a
+ * dev-ex improvement tool while working on tests.
+ *
+ * @return {Promise<void>}
+ */
```
Can we call teardown manually to accommodate for this need? Are we just clearing attributes and not everything else? Alternatively, can we add attributes in a more defensive way so an error won't be returned if attributes already exist? If that's not possible, can we call all teardown functions in the setup but guard that in a flag that's only true when working locally?

The issue I have with this is that this adds extra load to tests, and doesn't feel like it's enough, which might result in us adding more teardown calls in setup and hindering our E2E speed.

ralucaStan replied:
I've removed the `ensureCleanAttributes` function and created a new issue for handling the clearing of the attributes https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5410

senadir commented on tests/e2e/specs/backend/active-filters.test.js:
```diff
@@ -0,0 +1,106 @@
+	class: '.wc-block-active-filters',
+};
+
+if ( process.env.WP_VERSION < 5.3 )
+	// eslint-disable-next-line jest/no-focused-tests
+	test.only( 'skipping all other things', () => {} );
```
This can be deleted. We don't support 5.3 anymore.

senadir commented on tests/e2e/specs/backend/attribute-filter.test.js:
```diff
@@ -0,0 +1,176 @@
+	class: '.wc-block-attribute-filter',
+};
+
+if ( process.env.WP_VERSION < 5.3 )
+	// eslint-disable-next-line jest/no-focused-tests
+	test.only( 'skipping all other things', () => {} );
+
```
Same here.

senadir commented on tests/e2e/specs/backend/active-filters.test.js:
```diff
@@ -0,0 +1,106 @@
+			{ text: 'New Title' }
+		);
+
+		expect( await getEditedPostContent() ).toMatchSnapshot();
```
I think we can get rid of this given that we use fixtures for this block which already ensures that attributes persist.

senadir commented on tests/e2e/specs/backend/price-filter.test.js:
```diff
@@ -0,0 +1,129 @@
+				'.wp-block[data-type="woocommerce/price-filter"] textarea.wc-block-editor-components-title',
+				'Filter by price'
+			);
+			// Change title to h3.
+			const heading3Button = await queries.getByLabelText(
+				document,
+				/Heading 3/i
+			);
+			await heading3Button.click();
+			await expect(
+				page
+			).not.toMatchElement(
+				'.wp-block[data-type="woocommerce/price-filter"] h6 textarea',
+				{ text: 'New Title' }
+			);
+		} );
```
Resetting the block back to its initial state after will get painful quickly. Might be worth exploring this in a follow-up?
https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/2783

senadir commented on tests/e2e/specs/backend/price-filter.test.js:
```diff
@@ -0,0 +1,129 @@
+			await expect( page ).toMatchElement(
+				'button.wc-block-filter-submit-button.wc-block-price-filter__button'
+			);
+			expect( await getEditedPostContent() ).toMatchSnapshot();
```
As I said, block fixtures should solve the need for this.

senadir reviewed (commented):
Lovely test suit Thomas, Thank you so much for reviving #2756
I left some minor comments. I would differ the approval/decline to someone else.

opr committed:
Add `puppeteer-testing-library`

opr committed:
Use puppeteer-testing-library API to look for H6 button

opr committed:
Use correct argument in jest-puppeteer.dev.config

ralucaStan committed:
Update package.json & package-lock.json

ralucaStan committed:
remove tests for 5.3

ralucaStan committed:
remove snapshots

ralucaStan committed:
Remove ensureCleanAttributes

ralucaStan reviewed (dismissed):
I've rebased this to trunk to use node v16 and handled the requested changes.

mikejolley commented:
Removing RTM label and milestone until updated.

ralucaStan commented on tests/e2e/config/setup.js:
```diff
@@ -39,14 +39,19 @@ module.exports = async ( globalConfig ) => {
 			createShippingZones(),
+			createProductAttributes(),
 			enablePaymentGateways(),
-			setupPageSettings(),
```
@opr is there a reason this was removed?

opr replied:
I can't remember, can you add it back and does it still work fine if you do?

ralucaStan committed:
Remove puppeteer-testing-library, install missing expect-puppeteer package and update tests

The puppeteer-testing-library, although added by gutenberg is not actively used.
I have removed it and realised that the added functionality can be easily replaced by expect-puppeteer.
This is a package that we specify in the jest configuration but it was not part of our package.json.
The selectors from puppeteer-testing-library have been replaced with .toClick and the utility clearAndFillInput with .toFill

The Filter by Attribute block has been skiped until tests are fixed.

ralucaStan committed:
Add back setupPageSettings to jest setup

ralucaStan committed:
Fix filter by attribute block test

ralucaStan committed:
Fix linting error

ralucaStan commented:
The puppeteer-testing-library, although added by Gutenberg is not actively used.
I have removed it and realised that the added functionality can be easily replaced by `expect-puppeteer`, which Gutenberg also uses in its tests.
This is a package that we specify in the jest configuration but it was not part of our package.json.
The selectors from puppeteer-testing-library have been replaced with `.toClick` and the utility `clearAndFillInput` with `.toFill`

ralucaStan reviewed (approved):
I've picked this PR after it was opened for almost 4 months and made some changes:
The puppeteer-testing-library, although added by Gutenberg is not actively used.
I have removed it and realised that the added functionality can be easily replaced by expect-puppeteer, which Gutenberg also uses in its tests.
This is a package that we specify in the jest configuration but it was not part of our package.json.
The selectors from puppeteer-testing-library have been replaced with .toClick and the utility clearAndFillInput with .toFill
